### PR TITLE
Use per-derivation sandbox profiles

### PR DESCRIPTION
**Pull request by pikajude**
@copumpkin @henrytill @jwiegley @datakurre and anybody else interested in Darwin stuff.

This solution is in my opinion more viable long-term than continually adding more permissions to the default nix sandbox.

See https://github.com/NixOS/nix/pull/695

### CODE REVIEW

**Comment by datakurre** [comment]
As a user I'm positive for trying this out, but I recall @copumpkin being considered about some security implications.

File: pkgs/applications/version-management/git-and-tools/git/default.nix [hunk 1]
```diff
@@ -36,6 +36,9 @@ stdenv.mkDerivation {
   # required to support pthread_cancel()
   NIX_LDFLAGS = stdenv.lib.optionalString (!stdenv.isDarwin) "-lgcc_s";
 
+  # without this, git fails when trying to check for /etc/gitconfig existence
+  __propagatedSandboxProfile = stdenv.lib.sandbox.allowDirectoryList "/etc";
```
Comment (nbp) [inline pkgs/applications/version-management/git-and-tools/git/default.nix]: `__` prefixes are used for builtins.  I would recommend against this naming convention, and suggest that you use a single `_` prefix if such prefixes are needed.

**Comment by shlevy** [comment]
Merged to staging, this PR will close at next staging merge.

**Comment by shlevy** [comment]
Ah, didn't see @nbp's comment. Please fix and open a PR against staging when ready.

**Comment by shlevy** [comment]
@pikajude Isn't generate-framework-profile inherently impure? Why is that here and not part of the build hook? It being here will mean the derivation will be different if instantiated on different versions.

**Comment by pikajude** [comment]
@shlevy And it's specifically never substituted, which means that the derivation will have the same hash on host systems with different structures, but will always contain the dependencies that are specific to that system. I wanted to avoid a situation where people could have a copy of the profile substituted from Hydra that would cause link failures when building other packages.

I liked this solution better than using the pre-build-hook in Nix, though for all intents and purposes they're the same. Do you have a compelling argument for relegating the functionality back into Nix over keeping it here?

**Comment by shlevy** [comment]
I'm not sure what you mean by "it's never substituted". Derivations will have a different `__sandboxProfile` string on different systems, right? That will result in a different hash.

**Comment by pikajude** [comment]
No, how would it have changed? The definition of `generateFrameworkProfile` will be the same across systems, but the contents will vary.

**Comment by shlevy** [comment]
Ah, so you use stuff like `__propagatedSandboxProfile = stdenv.lib.sandbox.importProfile (generateFrameworkProfile name);`. That's really broken, if the result of generateFrameworkProfile is substituted (or inherited from an old version of NixOS, etc.), it won't work. We can't have impurities like that in nixpkgs.

**Comment by pikajude** [comment]
As I said, that's why I set `allowSubstitutes = false`.

**Comment by shlevy** [comment]
There's a huge difference between an impure path in the store and an impurity in the system itself. If I have a build farm set up, how can I guarantee that the machine used to build the profile is the same as the one used to build the resultant derivation? I can't. If `pre-build-hook` needs to be improved, let's improve it, if we need some different *principled* way of ensuring this impurity is localized, let's implement it, but doing it this way breaks nix.

**Comment by pikajude** [comment]
Ok, makes sense. The old `pre-build-hook` still exists and should be able to read __impureHostDeps, so should I just revert to that?

**Comment by shlevy** [comment]
For now, yeah, I think so (though not on this branch, as it's already merged into staging :smile: ). Can you maybe open up an issue on nix for what you would like the interface/error messages/etc. coming from a better mechanism might look like?

**Comment by pikajude** [comment]
@shlevy I can use `__propagatedImpureHostDeps` on frameworks only if you want, and keep sandbox profiles everywhere else.

**Comment by shlevy** [comment]
Cool, thanks

**Comment by pikajude** [comment]
@shlevy also @nbp proposed not using underscores at all for the sandbox profile attributes, how do you feel about that? It does make derivations look a bit cleaner.

**Comment by shlevy** [comment]
Fine by me.

**Comment by pikajude** [comment]
@shlevy: see https://github.com/NixOS/nixpkgs/pull/11193

**Comment by pikajude** [comment]
@edolstra btw, urllib3 DOES require network in its checkPhase, do you have any idea why it passes on Linux?

**Comment by shlevy** [comment]
@pikajude @edolstra I think Linux chroot enables localhost networking, right? Is that all urllib3 needs?

**Comment by copumpkin** [comment]
The OSX sandbox profile we were using enabled localhost networking too, so
that shouldn't break.

**Comment by ncfavier** [comment]
Does anyone remember why `/dev/ptmx` was needed for vim, and if it can be removed now?

**Pull request merged**